src/legacy/ui/public/capabilities/react/ui_capabilities_provider.tsx

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import PropTypes from 'prop-types';
```

Comment:
__issue:__ If we decide to go with React Context - we'll need to switch to a newer React Context API instead (`Context.Provider` & `Context.Consumer`). In I18n we have to use the Legacy API since `react-intl` is still not upgraded to latest context API and we just doesn't have any other choice.

I haven't touched this new API yet, but I'm going to experiment with this a bit and see what I can come up with.

Reply:
Converting this to the new context provider interfaces wasn't too difficult; however, it looks like the two approaches do not intermingle well. For the specific uses in the infra apps, the `withThemes` function provided by `styled-components` uses the old context syntax, so it was breaking the usage. 

I went ahead and committed the future version https://github.com/elastic/kibana/pull/31843/commits/28612d09512b8710117148021e607d8a78b984dd so we can switch to this version when we can.

Reply:
When I switched all usages within the infra app, the logging and infrastructure pages worked properly, but the metrics page broke with the following:

```
react-dom.development.js:506 Warning: WithTheme(InjectUICapabilities(InjectIntl(MetricDetailPage))) declares both contextTypes and contextType static properties. The legacy contextTypes property will be ignored.
warningWithoutStack @ react-dom.development.js:506
styled-components.browser.cjs.js:2201 [withTheme] You are not using a ThemeProvider nor passing a theme prop or a theme in defaultProps
componentWillMount @ styled-components.browser.cjs.js:2201
react-dom.development.js:506 Warning: Render methods should be a pure function of props and state; triggering nested component updates from render is not allowed. If necessary, trigger nested updates in componentDidUpdate.

Check the render method of MetricDetailPage.
warningWithoutStack @ react-dom.development.js:506
react-dom.development.js:506 Warning: Cannot update during an existing state transition (such as within `render`). Render methods should be a pure function of props and state.
warningWithoutStack @ react-dom.development.js:506
fatal_errors_service.tsx:36 Error: Uncaught TypeError: Cannot read property 'eui' of undefined (http://localhost:5601/bundles/infra.bundle.js:40107)
    at push../src/legacy/ui/public/notify/notify.js.window.onerror (notify.js:162)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:199)
    at invokeGuardedCallback (react-dom.development.js:256)
    at replayUnitOfWork (react-dom.development.js:18372)
    at renderRoot (react-dom.development.js:19262)
    at performWorkOnRoot (react-dom.development.js:20136)
    at performWork (react-dom.development.js:20048)
    at performSyncWork (react-dom.development.js:20022)
    at requestWork (react-dom.development.js:19891)
    at scheduleWork (react-dom.development.js:19705)
FatalErrorsService.add @ fatal_errors_service.tsx:36
host.ts:33 Uncaught TypeError: Cannot read property 'eui' of undefined
    at hostLayoutCreator (host.ts:33)
    at class_1../x-pack/plugins/infra/public/pages/metrics/index.tsx.class_1.render (index.tsx:58)
    at finishClassComponent (react-dom.development.js:14535)
    at updateClassComponent (react-dom.development.js:14490)
    at beginWork (react-dom.development.js:15438)
    at performUnitOfWork (react-dom.development.js:19106)
    at workLoop (react-dom.development.js:19146)
    at HTMLUnknownElement.callCallback (react-dom.development.js:149)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:199)
    at invokeGuardedCallback (react-dom.development.js:256)
fatal_errors_service.tsx:38 Uncaught Error: Uncaught TypeError: Cannot read property 'eui' of undefined (http://localhost:5601/bundles/infra.bundle.js:40107)
    at push../src/legacy/ui/public/notify/notify.js.window.onerror (notify.js:162)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:199)
    at invokeGuardedCallback (react-dom.development.js:256)
    at replayUnitOfWork (react-dom.development.js:18372)
    at renderRoot (react-dom.development.js:19262)
    at performWorkOnRoot (react-dom.development.js:20136)
    at performWork (react-dom.development.js:20048)
    at performSyncWork (react-dom.development.js:20022)
    at requestWork (react-dom.development.js:19891)
    at scheduleWork (react-dom.development.js:19705)
```

I was actually just able to get past this issue by reordering from `withThemes(injectUICapabilities(...))` to `injectUICapabilities(withThemes(...))`, so let's go with that for the time being. I'll switch the structure of the folders for `ui/capabilties/react` so the root is the new context api, and `ui/capabilities/react/legacy` is the old approach, just incase we can't take this solution elsewhere.